File: mint-runtime-permission/src/main/java/net/crystalapps/permission/runtime/library/config/RuntimePermissionConfig.java
package net.crystalapps.permission.runtime.library.config;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import net.crystalapps.permission.runtime.library.callbacks.PermanentDeniedHandler;
import net.crystalapps.permission.runtime.library.callbacks.PermissionFailureHandler;

@SuppressWarnings("unused")
public class RuntimePermissionConfig{

    private static RuntimePermissionConfig INSTANCE;

    @Nullable
    private PermanentDeniedHandler permanentDeniedHandler;

    @NonNull
    private final PermanentDeniedHandler defaultPermanentDeniedHandler;

    private PermissionFailureHandler permissionFailureHandler;

    private RuntimePermissionConfig() {
        defaultPermanentDeniedHandler = new MintPermanentDeniedHandler();
    }

    public static RuntimePermissionConfig getInstance() {
        return INSTANCE == null ? INSTANCE = new RuntimePermissionConfig() : INSTANCE;
    }

    public RuntimePermissionConfig setPermanentDeniedHandler(@Nullable PermanentDeniedHandler permanentDeniedHandler) {
        this.permanentDeniedHandler = permanentDeniedHandler;
        return this;
    }

    @NonNull
    public PermanentDeniedHandler getPermanentDeniedHandler() {
        return permanentDeniedHandler == null ? defaultPermanentDeniedHandler : permanentDeniedHandler;
    }

    public PermissionFailureHandler getPermissionFailureHandler() {
        return permissionFailureHandler;
    }

    public RuntimePermissionConfig setPermissionFailureHandler(@NonNull PermissionFailureHandler permissionFailureHandler) {
        this.permissionFailureHandler = permissionFailureHandler;
        return this;
    }
}

File: mint-runtime-permission/src/main/java/net/crystalapps/permission/runtime/library/config/RuntimePermissionConfig.kt
package net.crystalapps.permission.runtime.library.config

import net.crystalapps.permission.runtime.library.callbacks.PermanentDeniedHandler
import net.crystalapps.permission.runtime.library.callbacks.PermissionFailureHandler

class RuntimePermissionConfig private constructor() {

    private var permanentDeniedHandler: PermanentDeniedHandler? = null

    private val defaultPermanentDeniedHandler: PermanentDeniedHandler = MintPermanentDeniedHandler()

    private var permissionFailureHandler: PermissionFailureHandler? = null



    fun setPermanentDeniedHandler(permanentDeniedHandler: PermanentDeniedHandler?): RuntimePermissionConfig {
        this.permanentDeniedHandler = permanentDeniedHandler
        return this
    }

    fun getPermanentDeniedHandler(): PermanentDeniedHandler {
        return permanentDeniedHandler ?: defaultPermanentDeniedHandler
    }

    fun getPermissionFailureHandler(): PermissionFailureHandler? {
        return permissionFailureHandler
    }

    fun setPermissionFailureHandler(permissionFailureHandler: PermissionFailureHandler): RuntimePermissionConfig {
        this.permissionFailureHandler = permissionFailureHandler
        return this
    }

    companion object {

        private var INSTANCE: RuntimePermissionConfig? = null

        val instance: RuntimePermissionConfig
            get() {
                if (INSTANCE == null) INSTANCE = RuntimePermissionConfig()
                return INSTANCE!!
            }
    }
}